generation/day_ahead.py

```diff
@@ -0,0 +1,103 @@
+# You'll need a separate account & access token from ENTSO-E for this platform, though.
+# TODO: move to main __init__, and document it in Readme
+# entsoe.entsoe.URL = "https://iop-transparency.entsoe.eu/api"
+
+@entsoe_data_bp.cli.command("import-day-ahead-generation")
+@click.option(
+    "--dryrun/--no-dryrun",
+    default=False,
+    help="In Dry run, do not save the data to the db.",
+)
+@with_appcontext
+@task_with_status_report
+def import_day_ahead_generation(dryrun: bool=False):
+    """
+    Import forecasted generation for the upcoming day.
+    This will save overall generation, solar, offshore and onshore wind, and the estimated CO2 content per hour.
+    Possibly best to run this script somewhere around or maybe two or three hours after 13:00,
+    when prices are announced.
+    """
+    log = current_app.logger
+    country_code = current_app.config.get("ENTSOE_COUNTRY_CODE", DEFAULT_COUNTRY_CODE)
+    timezone = current_app.config.get("ENTSOE_TIMEZONE", DEFAULT_TIMEZONE)
+
+    log.info(f"Will contact ENSO-E at {entsoe.entsoe.URL}, country code: {country_code}, timezone {timezone} ...")
+
+    ensure_data_source()
+    ensure_generation_sensors()
+
+    now = server_now().astimezone(pytz.timezone(timezone)).replace(minute=0, second=0, microsecond=0)
+    from_time = (now + timedelta(hours=24)).replace(hour=0)
+    until_time = from_time + timedelta(hours=24)
+    log.info(
+        f"Importing generation data from ENTSO-E, starting at {from_time}, up until {until_time} ..."
+    )
+    from_time = from_time.astimezone(pytz.utc)
+    until_time = until_time.astimezone(pytz.utc)
+
+    auth_token = current_app.config.get("ENTSOE_AUTH_TOKEN")
+    if not auth_token:
+        click.echo("Setting ENTSOE_AUTH_TOKEN seems empty!")
+        raise click.Abort
+    
+    client = EntsoePandasClient(api_key=auth_token)
+    log.info("Getting all generation ...")
+    all_generation: pd.DataFrame= client.query_generation_forecast(country_code, start=from_time, end=until_time)
+    log.debug(all_generation)
+    log.info("Getting green generation ...")
+    green_generation_df: pd.DataFrame = client.query_wind_and_solar_forecast(country_code, start=from_time, end=until_time, psr_type=None)
+    log.debug(green_generation_df)
+
+    log.info("Down-sampling green energy forecast ...")
+    green_generation_df = green_generation_df.resample("60T").mean()    
+    log.debug(green_generation_df)
+
+    log.info("Aggregating green energy columns ...")
+    green_generation = green_generation_df.agg(["sum"], axis="columns")
+    log.debug(green_generation)
+
+    log.info("Computing grey generation forecast ...")
+    grey_generation = all_generation - green_generation["sum"]
+    log.debug(grey_generation)
+
+    # TODO: compute CO2 content from these MWh values
+    
+    # TODO: save values for each sensor we use, via fm.api.common.api_utils.save_to_db (make BeliefsDataFrames first)  
```

Comment:
Actually, that function gets a JSON structure as input, while I already have a `pd.DataFrame` (which I can turn into a BeliefsDataFrame with little effort). I would hhave to build something Marshmallow can parse so validations can work ...

The checks we're missing right now are the unit and the resolution (which also can be upsampled if possible).

How about I add to `flexmeasures.api.common.schemas.sensor_data` a function which can apply the same validations on a `BeliefsDataFrame`? Maybe I'm overlooking a simpler solution, but at least we'd keep things in the same place.

Reply:
> [...] a function which can apply the same validations on a BeliefsDataFrame [...]

I fear it may not be so easy to use Marshmallow validators on a DataFrame. [Here's a related discussion.](https://github.com/marshmallow-code/marshmallow/issues/1181) I think you may need to choose between converting the DataFrame to something else (dict or JSON) first, or not use Marshmallow for the validation.

The latter doesn't appeal to me, seeing that we set out to unify how we handle validation, rather than diversify.

The former feels kind of similar to getting your DataFrame into a JSON that can be send to the FM API, right? Having such a util function—one that plugins can use to transform a simple pandas time Series into a FM API compatible JSON—sounds pretty useful to me.

Reply:
There must be a better way than getting XML from an API turned into a DataFrame, then turning that into JSON, and then back into a BeliefsDataFrame to finally get it into an SQLAlchemy ORM model wich creates SQL statements.

It's crazy as it is, but the JSON step is the icing on the cake. My train of thought atm is tthat we'd keep the validations in one module at least, to have an overview in one place, where the two main ways to validate are.